### DISCUSSION TITLE
Metrics Explorer v1.0 Specifications

### DISCUSSION THREAD

**Discussion by celian-garcia** [perses/perses#1859] [Design Discussions]
Hello, I would like to propose some improvement axes to bring to the current metrics explorer. 
Creating this discussion to study a bit of design and collect feedbacks before really starting implementation. 

I see several main categories 
- **Query Result** - the way we display the result of a PromQL query
- **Explore** - the way we provide info to the user of which metrics, labels, value are available - **REMAIN THE MAIN PURPOSE**
- **Ergonomy** - of course, it should be as easy as possible for the user to navigate.
- **Technical architecture** - specially regarding dependencies. code files should remain independent from ``dashboard`` code

## Query Result
We have range query result.
I see two possibilities to add
### **instant** query result:
A table to display only the T time timeseries, like in Prometheus/Thanos/Promlens
![image](https://github.com/perses/perses/assets/10258608/b2dcaf8e-ae69-4eea-aaa9-3fbda0cf8e00)

### **explain** tab like promlens. I have no idea of the effort it would represent
![image](https://github.com/perses/perses/assets/10258608/cdd3aad8-9cb5-4bcd-a18b-185000cc9507)

## Explore
We already have a quite significant amount of information 
![image](https://github.com/perses/perses/assets/10258608/608de691-8de6-496b-be73-a36326a36241)

But this can be more. Again checking promlens capabilities. We could have also labels availables, with their values, search capabilities
![image](https://github.com/perses/perses/assets/10258608/0e46580b-7206-4ecd-9eda-4d89fd09ce67)

> Note that performances are a tricky topic as we already know that this can be hard to retrieve this detail of information in companies managing a lot of prometheus and a lot of clusters like in Amadeus.
> For this particular type of issue, we can maybe inspire from some Amadeus tricks: here we propose to user type the start of their metric to make the exploring query only on one metric at at time
> ![image](https://github.com/perses/perses/assets/10258608/ea8eeb5b-fcfa-4268-990d-5b67ec7e8ae6)

## Ergonomy & Technical architecture
Nothing to share here. Everything is in the slight desc on top.

**Comment by nicolastakashi** [comment]
Thanks for handling this mate!
I'll share my initial insights about this.

1. First of all I do believe we should start by decoupling the Explore Page and the Dashboard Page, initially the idea was to avoid as much code duplication as possible, since in theory the explore page is just a static dashboard.
But after we have this working for a while, we're pretty sure the even this looking similar the pages are super different, especially because the Dashboard Page is very optioned on how the components will be rendered, and I do think we may want a different UX/UI for the explore page.

2. After we decouple this a missing feature is the Instant Query and the Table View, I remember when I started to join the project @sjcobb mentioning about some refactory we need to do in other to be able to ensure the right query type will be selected based on the panel we're using, for example if I'm selecting a table panel this means the only query type available is instant, but if I'm selecting a time series chart I might be able to execute a instant or a range query.

Tbh this sounds good but very hard to implement I don't know if we should tie the instant query with this functionality, I'd say we are good to move forward without this logic and offer instant query as a second feature.

3. Regarding Metrics Explore maybe I'm not the best one because I like to do this in a very adhoc manner, by typing metrics names and just reading and exploring it.

**Reply by celian-garcia** [reply]
Great thanks you for the details! It will help me for sure. I think we are on the same page that's cool. 

I like the fact that you place on the Frontline the decoupling with dashboard. This is also a priority to me. 
I'm just maybe overconfident on the instant query feature but we'll see. Instant query and table are specially two feature that we'd like to tackle in the coming weeks.

**Reply by nicolastakashi** [reply]
I'm more than available to support reviewing and brainstorming

**Reply by sjcobb** [reply]
With instant queries, the idea was to add `supportedQueryInterfaces` to each plugin where they are registered so we could define whether they work with a single datapoint or not, table actually should support all query types. Instant query would also work with StatChart (without sparkline) but not TimeSeriesChart. So with this viz-first approach it’d be easier for users to tell which visualization works with which query type instead of always showing the full list of plugins, also certain options like `calculation` would be disabled when we know only a single datapoint is returned.

This approach would make things harder on plugin developers since some panel plugins would need to work with multiple query types and there is a lot of [refactoring needed](https://github.com/perses/perses/issues/1303#issuecomment-1648332353) in DataQueriesProvider. It also is maybe overkill if we aren’t going to have that many core panel plugins, doing it similar to Grafana would definitely be easier to implement and migrate from their schema

**Reply by celian-garcia** [reply]
I'm realising that I didn't answer this message. Some work as already be done since but basically the only thing I'd say to complement is that from the experience I have with the explorer, the instant query should not be a separate query type.
In other words I'm not in favour of having the query type selector saying 
- Prometheus Range Query 
- Prometheus Instant Query 

As it is the same query but just executed and displayed differently.

So at the end. We'd make something more like Grafana, where the selection is only 
- Prometheus Query

and an additional selector at panel view would allow you to switch from instant to range. 

For now I implemented it without the switch here 
https://github.com/perses/perses/pull/1982 (the panel plugin hardly defines if the queries are all instant or all range)

But then we might think of having this switch and save in the panel spec the value. It's out of the explorer scope.

**Comment by celian-garcia** [comment]
As I leaked it in a pull request, I can put it there. Work in progress design. This is a naive design to trigger discussion. Don´t hesitate to provide feedback. 
The big deal will be to choose if and how we integrate the prometheus explorer view INTO the prometheus plugin.
My opinion is that we should! But the how is probably not that simple.

![image](https://github.com/perses/perses/assets/10258608/ecdb4696-c5aa-4fd5-9181-e20fe8ee1fba)

**Comment by Nexucis** [comment]
Today the explorer feature has been fragmented with the new plugin system. It means, a datasource can provide its own explorer now.

As the changes related to this new architecture has been implemented and will be released soon, I think we can close this discussion.

Design discussion or issue, should depend on the plugin itself now.
